Amend nosleep command to accept awake time

clumsy-stefan opened the pull request:
Amend system command "nosleep" to accept a parameter that sets the awake time (eg. nosleep for <time>).
like this it's possible to send a "nosleep" command to the unit immediately after it awakes, do whatever needed and then reset the awake time with "nosleep,<time>" instead of having to go via web-page.

clumsy-stefan committed:
Amend nosleep command to accept awake time

Amend system command "nosleep" to accept a parameter that sets the awake time (eg. nosleep for <time>).
like this it's possible to send a "nosleep" command to the unit immediately after it awakes, do whatever needed and then reset the awake time with "nosleep,<time>" instead of having to go via web-page.

clumsy-stefan committed:
Merge branch 'mega' into amend_system_nosleep

Grovkillen commented:
Should nosleep render an event? `System#NoSleep` in order to publish something or light an LED or whatever to indicate that it has entered that mode?

clumsy-stefan commented:
@Grovkillen sorry, not sure if I understand your question?

the change is really just to be able to remotely set the awake time (quickly) and automatically (from a script or controller) for units that are usually in deepsleep and only awaken for a short time. so basically to intercept the wakeup from remote).

Grovkillen commented:
Yes for sure but you might need a script to get an response from the unit that it has entered the NoSleep mode and want to use rules to do things that is only related to that mode being entered. We want to have event for more things to be able to create better rule handling.

clumsy-stefan commented:
I guess that would be a good thing, but I have to admit, I don't know how to do that correctly...

Until now when it entered that mode (after changeing it on the web page) was there an even triggered?

Grovkillen commented:
No it is no such event currently but we want it :)

clumsy-stefan commented:
ahh... I think I understand... sorry, I'm sometimes a bit slow in understanding...

as said, I'd need to look into that on how this is done, if you want to include it in this PR...

Grovkillen commented:
Yes please do :+1:

clumsy-stefan commented:
I think I found the right place.... testing now ;)

clumsy-stefan committed:
Add System#NoSleep=<time> event

Add a System#NoSleep event which returns the amount of seconds fow which the unit will be awake (at least). However this time can change within the delay or even disabled (by rules, commands, etc.)

clumsy-stefan commented:
Addes a System#NoSleep event which returns the amount of seconds for which the unit will be awake (at least). However this time can change within the delay or even disabled (by rules, commands, etc.)

TD-er commented on src/ESPEasy.ino:
```diff
@@ -248,6 +248,12 @@ void setup()
+  {
+    String event = F("System#NoSleep=");
+    event += Settings.deepSleep;
+    rulesProcessing(event);
```
Is the `deepSleep` setting actually the "awake" time?
If so, then we have to change it, since it is very confusing.
I have to look at it when I will be near my computer again.

clumsy-stefan replied:
as far as I understand yes. in Misc.ino (~line 1155):
```
  addFormNumericBox( F("Sleep awake time"), F("deepsleep"), Settings.deepSleep, 0, 255);
  addUnit(F("sec"));
  addHelpButton(F("SleepMode"));
  addFormNote(F("0 = Sleep Disabled, else time awake from sleep"));
  addFormNumericBox( F("Sleep time"), F("delay"), Settings.Delay, 0, 4294);   //limited by hardware to ~1.2h
  addUnit(F("sec"));
```
`Settings.deepSleep` is the actual awake time and `Settings.Delay` is the time to sleep..  bit confusing, the other way round would probably be more self speaking.

I think this iwas already discussed in another thread (also how it should be called in the web-interface)...

So it could/should probably be refactored everywhere.

TD-er replied:
Yep, refactoring should be another issue/commit.
And at the same time make it more comprehensive to people.
Every now and then there is another issue/question about it, so it is not intuitive apparently.

clumsy-stefan commented:
these changes/PR runs fine for a couple of hours now on all my nodes, some of them actually doing a deepsleep (running on battery), sleeping for 10min. and waking for 15sek. so I from my point of view it could be merged, if desired.
